### PR TITLE
[RW-281] Replacing register endpoint with endpoints for submitting TOS consent,

### PR DESCRIPTION
**Pull request by danqrodney**
ethics training, demographics survey response.

Changing ID verification endpoint to return profile (which may have its data access level updated as a result of ID verification.)

### CODE REVIEW

**Comment by danqrodney** [comment]
Adding tests now; sending out for feedback on the API. (Which is sort of a placeholder at this point, since we don't have details for how TOS, ethics training, and demographic surveys will look.)

File: api/src/main/resources/workbench.yaml [hunk 1]
```diff
@@ -141,38 +141,61 @@ paths:
...
           description: Account deleted successfully.
 
-  /api/v1/register:
+  /api/v1/id-verification:
     post:
       tags:
         - profile
-      description: Registers a profile for the user in AllOfUs and Firecloud; called after the user has consented to the terms of use
-      operationId: register
+      description: Accepts identity information for verification.
```
Comment (calbach) [inline api/src/main/resources/workbench.yaml]: How do these three methods relate to one another? Do they need to be called in a particular order? Does something happen once all of them have been called?

File: api/src/main/resources/workbench.yaml [hunk 2]
```diff
@@ -141,38 +141,61 @@ paths:
         "204":
           description: Account deleted successfully.
 
-  /api/v1/register:
+  /api/v1/id-verification:
```
Comment (calbach) [inline api/src/main/resources/workbench.yaml]: Was there no client code that called the previous "register" endpoint? Is it still possible to create a researcher account locally and get into a valid state? Or is the registration / verification stuff just not yet considered for accessing the frontend?

Comment (calbach) [inline api/src/main/resources/workbench.yaml]: nit: existing issue, but I'd generally recommend naming custom REST actions with verb names, e.g. verify-id, accept-terms-of-service

File: api/src/main/java/org/pmiops/workbench/db/dao/UserService.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
+package org.pmiops.workbench.db.dao;
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: WorkspaceService follows a interface/impl pattern. Is there a reason to deviate from that pattern here?

File: api/src/main/java/org/pmiops/workbench/db/dao/UserService.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      @Override
+      public User apply(User user) {
+        user.setBlockscoreId(blockscoreId);
+        user.setBlockscoreVerificationIsValid(blockscoreVerificationIsValid);
+        return user;
+      }
+    });
+  }
+
+  public User submitTermsOfService() {
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: nit: feels slightly better to me to pass a timestamp and bind to the closure here. Feels a bit too arbitrary to recompute a different timestamp depending on how many times the mutation was retried.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserService.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    this.clock = clock;
+  }
+
+  /**
+   * Updates a user record with a modifier function.
+   *
+   * Ensures that the data access level for the user reflects the state of other fields on the
+   * user; handles conflicts with concurrent updates by retrying.
+   */
+  User updateWithRetries(Function<User, User> userModifier) {
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: private? Or `VisibleForTesting`?

File: api/src/main/java/org/pmiops/workbench/db/model/User.java [hunk 6]
```diff
@@ -170,4 +173,31 @@ public Boolean getBlockscoreVerificationIsValid() {
   public void setBlockscoreVerificationIsValid(Boolean value) {
     blockscoreVerificationIsValid = value;
   }
+
+  @Column(name = "terms_of_service_completion_time")
+  public Timestamp getTermsOfServiceCompletionTime() {
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/db/model/User.java]: Should these things be populated on the `Profile` object in the swagger API definition? Is there some other way for the UI to infer which of these things have been completed on the client?

Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/db/model/User.java]: Yeah, good point! Done.

File: api/src/main/resources/workbench.yaml (hunk 2 above)
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: The UI wasn't calling this yet AFAIK. That comes next, with these new endpoints.

Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: Done.

File: api/src/main/resources/workbench.yaml (hunk 1 above)
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: At the moment I'm assuming they could be called in any order -- the user chooses which steps to do when.

Once the user has completed ID verification, TOS, ethics training, and the demographic survey, their data access level switches to REGISTERED, and (this isn't implemented yet) we add them to the authorization domain for access to the CDR.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserService.java (hunk 3 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: I primarily care about consistency so things are unambiguous next time someone needs to add a `FooService`. I'd be fine with the latter suggestion (of merging workspace service / impl) if you don't see any value in splitting them. Doesn't necessarily need to be in this PR.

File: api/src/main/java/org/pmiops/workbench/api/ProfileController.java [hunk 7]
```diff
@@ -353,11 +334,25 @@ private User initializeUserIfNeeded() {
       request.getDocumentType(), request.getDocumentNumber()
     );
 
-    User user = userProvider.get();
-    user.setBlockscoreId(person.getId());
-    user.setBlockscoreVerificationIsValid(person.isValid());
-    userDao.save(user);
+    User user = userService.setIdVerification(person.getId(), person.isValid());
```
Comment (jmandel) [inline api/src/main/java/org/pmiops/workbench/api/ProfileController.java]: Do we want/expect to re-run this even after a successful verification has already been performed?

File: api/src/main/java/org/pmiops/workbench/db/dao/UserService.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    }
+  }
+
+  private void updateDataAccessLevel(User user) {
+    if (user.getDataAccessLevel() == DataAccessLevel.UNREGISTERED) {
+      if (user.getBlockscoreVerificationIsValid()
+        && user.getDemographicSurveyCompletionTime() != null
+        && user.getEthicsTrainingCompletionTime() != null
+        && user.getTermsOfServiceCompletionTime() != null) {
+        user.setDataAccessLevel(DataAccessLevel.REGISTERED);
```
Comment (jmandel) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: At some stage (follow-on work?) we want to make sure that this `DataAccessLevel.REGISTERED` update also triggers adding the user to the appropriate FireCloud authorization domain.

File: api/src/main/resources/workbench.yaml (hunk 1 above)
Comment (jmandel) [reply api/src/main/resources/workbench.yaml]: agreed we want to allow these to occur in any order! (The UI may guide the user toward some specific order...)

File: api/src/main/java/org/pmiops/workbench/api/ProfileController.java (hunk 7 above)
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ProfileController.java]: I'm not sure what the right UX is... it's not clear we want users to be able to attempt to re-verify themselves once they have done so successfully.

Anyway, at the moment if you call this method multiple times the last one wins.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserService.java (hunk 3 above)
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: OK. I'll merge them in a follow-up.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserService.java (hunk 8 above)
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: Yep, @blrubenstein is working on that integration.

**Comment by danqrodney** [comment]
Tests added.

File: api/src/main/resources/workbench.yaml [hunk 9]
```diff
@@ -141,38 +141,61 @@ paths:
...
       responses:
-        "204":
-          description: Request received and validated.
+        "200":
+          description: The user's profile.
+          schema:
+            $ref: "#/definitions/Profile"
+
+  # TODO: add demographic survey response data
+  /api/v1/account/submit-demographic-survey:
```
Comment (dmohs) [inline api/src/main/resources/workbench.yaml]: I have a naming suggestion. The verbs accept, complete, and submit imply some work, but I don't see that in the code itself. For clarity, based on reading the code these endpoints call, I would name them, "mark-ethics-training-completed", "mark-terms-of-service-accepted", etc.

File: api/db/changelog/db.changelog-16.xml [hunk 10]
```diff
@@ -13,3 +13,4 @@
 
     </changeSet>
 </databaseChangeLog>
+
```
Comment (dmohs) [inline api/db/changelog/db.changelog-16.xml]: Nit: May want to revert this file to keep git blame clean.

File: api/src/main/java/org/pmiops/workbench/api/ProfileController.java [hunk 11]
```diff
@@ -353,11 +334,25 @@ private User initializeUserIfNeeded() {
...
+    return getProfileResponse(user);
+  }
 
-    return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+  @Override
+  public ResponseEntity<Profile> submitDemographicsSurvey() {
+    User user = userService.submitDemographicSurvey();
```
Comment (dmohs) [inline api/src/main/java/org/pmiops/workbench/api/ProfileController.java]: To follow-on to my naming suggestion (below), I would expect a function with the verb "submit" to accept parameters. What am I submitting?

File: api/src/main/resources/workbench.yaml (hunk 9 above)
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: The demographics survey will contain answers to questions about the researcher's demographics (race, gender, etc...) this method currently is just a placeholder. So "submit" makes sense here, for ID verification, too.

Foraccepting terms of service, at a minimum I expect the user to check a box saying "I accept" or hit an "Accept" button. "Submit" still sounds appropriate there.

For ethics training, I'm not sure -- that might just be watching a video without any explicit submit action on the part of the user -- but we'll see. 

Personally, I'm OK with "submit" for all of this.

Comment (dmohs) [reply api/src/main/resources/workbench.yaml]: Okay, I respect your judgement.

File: api/src/main/java/org/pmiops/workbench/api/ProfileController.java (hunk 11 above)
Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/ProfileController.java]: See other comments.

File: api/src/main/resources/workbench.yaml (hunk 9 above)
Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: Changed to "completeEthicsTraining".

**Pull request merged**